File: Data Structures & Algorithms/design-hashset/submission-2.py
class ListNode:
    def __init__(self, val: int):
        self.val = val
        self.next = None

class MyHashSet:

    def __init__(self):
        self.arr = [None] * 10_000
        

    def add(self, key: int) -> None:
        k = key % len(self.arr)
        if self.arr[k] is None:
            self.arr[k] = ListNode(key)
        else:
            prev = None
            curr = self.arr[k]
            while curr:
                if curr.val == key:
                    return
                prev = curr
                curr = curr.next
            prev.next = ListNode(key)
        

    def remove(self, key: int) -> None:
        k = key % len(self.arr)
        if self.arr[k] is None:
            return
        curr = self.arr[k]
        if curr.val == key:
            self.arr[k] = curr.next
            return
        prev = curr
        curr = curr.next
        while curr:
            if curr.val == key:
                prev.next = curr.next
                return
            prev = curr
            curr = curr.next


    def contains(self, key: int) -> bool:
        k = key % len(self.arr)
        if self.arr[k] is None:
            return False
        curr = self.arr[k]
        while curr:
            if curr.val == key:
                return True
            curr = curr.next
        return False
        


# Your MyHashSet object will be instantiated and called as such:
# obj = MyHashSet()
# obj.add(key)
# obj.remove(key)
# param_3 = obj.contains(key)

File: Data Structures & Algorithms/design-hashmap/submission-2.go
type ListNode struct {
    key int
    val int
    next *ListNode
}

type MyHashMap struct {
    data []*ListNode
}

const sizeCap int = 1_000

func Constructor() MyHashMap {
    arr := make([]*ListNode, sizeCap)
    for i := range arr {
        arr[i] = &ListNode{key: -1, val: -1}
    }
    return MyHashMap{ data: arr }
}

func (this *MyHashMap) hash(key int) int {
    return key % sizeCap
}

func (this *MyHashMap) Put(key int, value int) {
    idx := this.hash(key)
    listNode := this.data[idx]
    for listNode.next != nil {
        listNode = listNode.next
        if listNode.key == key {
            listNode.val = value
            return
        }
    }
    listNode.next = &ListNode{key: key, val: value}
}

func (this *MyHashMap) Get(key int) int {
    idx := this.hash(key)
    listNode := this.data[idx]
    for listNode.next != nil {
        if listNode.next.key == key {
            return listNode.next.val
        }
        listNode = listNode.next
    }
    return -1
}

func (this *MyHashMap) Remove(key int) {
    idx := this.hash(key)
    listNode := this.data[idx]
    for listNode.next != nil && listNode.next.key != key {
        listNode = listNode.next
    }
    if listNode.next != nil {
        listNode.next = listNode.next.next
    }
}

/**
 * Your MyHashMap object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Put(key,value);
 * param_2 := obj.Get(key);
 * obj.Remove(key);
 */